fdbclient/include/fdbclient/CommitTransaction.h

```diff
@@ -123,59 +126,107 @@ struct MutationRef {
+	// Set checksum bit with mutation type
+	uint8_t typeOnChecksumPayload() const { return this->type | CHECKSUM_PAYLOAD_FLAG_MASK; }
+
+	// Return true if the mutation type indicating the mutation param carrying checksum
+	bool withChecksumPayload() const { return this->type & CHECKSUM_PAYLOAD_FLAG_MASK; }
+
+	bool isAtomicOp() const { return (ATOMIC_MASK & (1 << type)) != 0; }
+	bool isValid() const { return type < MAX_ATOMIC_OP; }
+
+	// If mutation checksum is enabled, we must set accumulative index before serialization
+	// Once accumulative index is set, it cannot change over time
+	void setAccumulativeChecksumIndex(uint16_t index) {
+		if (this->accumulativeChecksumIndex.present()) {
+			ASSERT(this->accumulativeChecksumIndex.get() == index);
+		} else {
+			this->accumulativeChecksumIndex = index;
+		}
+		return;
+	}
+
+	// Convert to normal mutation without payload and
+	// setup this->checksum and this->accumulativeChecksumIndex
+	void removeChecksumPayload() {
 		this->checksum.reset();
-		if (!withChecksum()) {
-			return Optional<uint32_t>();
+		this->accumulativeChecksumIndex.reset();
+		if (!this->withChecksumPayload()) {
+			return;
 		}
-		ASSERT(this->param2.size() >= 4);
-		const int idx = this->param2.size() - 4;
-		const uint32_t pc = *(const uint32_t*)(this->param2.substr(idx).begin());
-		this->type &= ~CHECKSUM_FLAG_MASK;
-		this->param2 = this->param2.substr(0, idx);
-		return pc;
+		this->type &= ~CHECKSUM_PAYLOAD_FLAG_MASK;
+		ASSERT(this->param2.size() >= 6); // 4 bytes for checksum, 2 bytes for accumulative index
+		int index = this->param2.size() - 6;
+		this->param2 = this->param2.substr(0, index);
+		this->checksum = *(const uint32_t*)(this->param2.substr(index, 4).begin());
+		index += 4;
+		this->accumulativeChecksumIndex = *(const uint16_t*)(this->param2.substr(index, 2).begin());
 	}
-	bool withChecksum() const { return this->type & CHECKSUM_FLAG_MASK; }
 
-	bool isAtomicOp() const { return (ATOMIC_MASK & (1 << type)) != 0; }
-	bool isValid() const { return type < MAX_ATOMIC_OP; }
+	// Attach 32-bit crc and 16-bit accumulative index to param2
+	Standalone<StringRef> withPayloadSuffix(Standalone<StringRef> param) {
+		ASSERT(this->checksum.present() && this->accumulativeChecksumIndex.present());
+		uint32_t cs = this->checksum.get();
+		uint16_t acsIdx = this->accumulativeChecksumIndex.get();
+		return param.withSuffix(StringRef((uint8_t*)&cs, 4)).withSuffix(StringRef((uint8_t*)&acsIdx, 2));
+	}
 
-	uint32_t populateChecksum() {
-		ASSERT(!this->withChecksum());
-		uint32_t c = crc32c_append(static_cast<uint32_t>(this->type), param1.begin(), param1.size());
-		crc32c_append(c, param2.begin(), param2.size());
+	// Generate 32 bits checksum and set it to this->checksum
+	// If this->checksum exists, compare the new one with the existing one. Trace error if mismatch
+	void populateChecksum() {
+		ASSERT(!this->withChecksumPayload());
+		ASSERT(this->accumulativeChecksumIndex.present());
```

Comment:
What's the purpose of this assert? `accumulativeChecksumIndex.present()`. Why an index is required for a `regular` mutation?